### ISSUE TITLE
Rulesets not working with nested validators

### ISSUE THREAD

**Issue by waynebrantley** [JeremySkinner/FluentValidation#194]
```
    public class SharedValidator : AbstractValidator<ISomeInterface>
    {
        public SharedValidator()
        {
            RuleSet("SomeRuleset", () =>
            {
                RuleFor(x => x.SomeField).NotNull().Length(1, 50);
            });
        }
    }

````
````
    public class MainValidator : AbstractValidator<ISomeInterface>
    {
        public MainValidator ()
        {
                RuleFor(x => x.TestField).NotNull();
                RuleFor(x => x).SetValidator(new SharedValidator ());
        }
    }
```

When you call the MainValidator with something like:

```
var main = new MainValidator ();
var validationResult = main .Validate(sharedItem, ruleSet: "SomeRuleset");
```

The SharedValidator is not run.  Of course if I do:

```
var main = new MainValidator ();
var validationResult = main .Validate(sharedItem, ruleSet: "SomeRuleset, "default");
//or
var validationResult = main .Validate(sharedItem);
```
The SharedValidator code works as normal.
Thoughts?

**Comment by JeremySkinner** [comment]
That's actually the correct behaviour...the MainValidator doesn't have a "SomeRuleset" defined, therefore no rules are selected for execution.

**Comment by waynebrantley** [comment]
As you may recall, I needed validators to work with interfaces and you helped with some mods for that.
The project has lots of classes each with a validator - normal stuff.
The main validator for the class is normally just a composed set of other pre-existing validators, usinng SetValidator().

Each of these use an interface that is common, so I am able to share a set of 15 validators with 100s of classes - where each class simply composes what it needs and adds any special concerns of that specific class implementation of the common interface.  Many times, my set of validators also have other validators inside them.   This makes it easy to add common validation with a single line of code - or pick and choose.   The end result is the validators for classes implementing this common interface is just one or two 'SetValidator' lines - where each of those validators contain several 'SetValidators'.   If the validation varies too much, you can just include the pieces you need and write the custom ones.

Everything works great.....and the system is slick....except this.

During the common process that uses these validators, there is 'prework' it needs to do and it needs a subset of the validators for the class.   Later on it runs all the validators.   So, in some of my bottom level validators I have rulesets defined so they can be hand selected.   This leads me to this very stuck point...

Sounds like it is a complex solution from your side.   I understand that the main validator looks for a ruleset to match - does not find one - and just exits.   What would be great is if it could also look at any SetValidators for rulesets also.

Any possibility of this enhancement?   Or even if you point me to the code, I could look at implementing if the design/concept is something you would agree with?

**Comment by JeremySkinner** [comment]
I'm not keen on this, and I don't think it's a change I'd want to make in the core (explanation below), *but* it is something that you should be able to implement with the existing extensibility points if you need this behaviour. 

The point of rulesets is to group rules together for selection. If that ruleset is not present, and no rules match, then that should be the end of the execution run. A child validator is no different from any other rule...so if it isn't included, then none of its children should be included either. Additionally, it would be a very significant, yet sutble, breaking change. So that's why I'm not going to make the change, but here's how you can do it anyway: :) 

The logic for selecting which rules to execute is controlled by IValidatorSelector implementations. This is exposed on the ValidationContext as the Selector property. The default implementation (MemberNameValidatorSelector) is used to filter by property (that is, if you invoke Validate(instance, "Surname", "ForeName") then the MemberNameValidatorSelector is what handles this). If you don't specify any properties, then the MemberNameValidatorSelector just includes all rules.

If you invoke the Validate overload that has a ruleset, then the RulesetValidatorSelector is used instead. Here's the code for it:

```
public class RulesetValidatorSelector : IValidatorSelector {
  readonly string[] rulesetsToExecute;

  public RulesetValidatorSelector(params string[] rulesetsToExecute) {
    this.rulesetsToExecute = rulesetsToExecute;
  }

  public bool CanExecute(IValidationRule rule, string propertyPath, ValidationContext context) {
    if (string.IsNullOrEmpty(rule.RuleSet) && rulesetsToExecute.Length == 0) return true;
    if (string.IsNullOrEmpty(rule.RuleSet) && rulesetsToExecute.Length > 0 && rulesetsToExecute.Contains("default", StringComparer.OrdinalIgnoreCase)) return true;
    if (!string.IsNullOrEmpty(rule.RuleSet) && rulesetsToExecute.Length > 0 && rulesetsToExecute.Contains(rule.RuleSet)) return true;
    if (rulesetsToExecute.Contains("*")) return true;

    return false;
  }
}
```

In your case, you can provide an implementation that explicitly checks if a child validator is being run, and include it for execution. Here's a working example:

```
public class IncludeChildValidatorSelector : IValidatorSelector {
	private IValidatorSelector _innerSelector;
	string[] _rulesetsToExecute;

	public IncludeChildValidatorSelector(params string[] ruleSets) {
		_rulesetsToExecute = ruleSets;
		_innerSelector = new RulesetValidatorSelector(ruleSets);
	}

	public bool CanExecute(IValidationRule rule, string propertyPath, ValidationContext context) {
		if (string.IsNullOrEmpty(rule.RuleSet) && _rulesetsToExecute.Length > 0) {
			bool isChildValidator = rule.Validators.FirstOrDefault() is ChildValidatorAdaptor;
			if (isChildValidator) {
				return true;
			}
		}

		return _innerSelector.CanExecute(rule, propertyPath, context);
	}
}
```

So this is a custom selector that will check if the current rule contains a child validator, and that validator has no ruleset defined. If that's the case, then it will always include it for execution.

This is the simple case...if you've called .When after the SetValidator, then you'd need to check if the validator has been wrapped by a DelegatingValidator. Also, if you want the same behaviour for child collection validators, then you'd need an additional check.

You can then use it when you call Validate:

`validator.Validate(new Person(), selector: new IncludeChildValidatorSelector("MyRuleset"));`

Here's a full test example:

```
[Fact]
public void Ruleset_selection_should_cascade_downwards_with_custom_selector() {
	var validator = new TestValidator3();
	var result = validator.Validate(new Person(), selector: new IncludeChildValidatorSelector("Names"));
	result.IsValid.ShouldBeFalse();
}
```
```
private class TestValidator2 : AbstractValidator<Person> {
	public TestValidator2() {
		RuleSet("Names", () => {
			RuleFor(x => x.Surname).NotNull().WithMessage("foo");
		});
	}
}

public class TestValidator3 : AbstractValidator<Person> {
	public TestValidator3() {
		RuleFor(x => x).SetValidator(new TestValidator2());
	}
}
```

I hope that makes sense and is useful...please let me know if you'd like me to clarify anything.

**Comment by waynebrantley** [comment]
In my case, I am calling .Validate directly, so I could do it without the 'global default'.
Questions.

If the rule contained a validator and that validator contained a rule that contained a validator and so on, will the above work?

Additionally, the difference in our 'thought' process is am treating the 'SetValidator()' as an additional 'include' of rules.  Where you treat it as 'just another validator'.  So, how about a compromise so that it is not a breaking change.  (I know I am biased, but I see this as the way it should work. :-) ).  Anyway, here are my thoughts.

1)  When I use it as a true validator (The way it was originally designed), it looks like `RuleFor(x=>x.SomeProperty).SetValidator(new SomeValidator())`

2)  Recently you adopted support for 'model validators'.  Things that could validate at the model level, and not a property level.  When I use it like that, I think of it more like an 'include'...so it would just add these rules to this validator type of thing.  This looks like ``RuleFor(x=>x).SetValidator(new SomeValidator())`

My proposal would be one of two things (to the core project).

1)  If the style is a 'model level validator' then it DOES use the rulesets as I proposed.
or
2)  Add a new way to add a validator so that it makes it clear.  This would be non-breaking, add additional feature the the library in such a way that it would be trivial to use.  This would look like ``RuleFor(x=>x).IncludeValidator(new SomeValidator())`

I guess what makes this a paint point now and it was not in the future was the addition of Interface support for validators and the 'model support' on the RuleFor.   These to sort of make this feature something that is now applicable, where it was not in the past when SetValidator() was really meant as a way to quickly write some custom code to validate a particular field when it was complex.  

First option is slick, but being slick it is not specific and might be considered breaking.  (But only breaking to those who have added a SetValidator() to a model level rule.  And that has not been available that long)    Second option is not slick and very specific opt-in feature, so that might be the way to do it.

Thanks for listening!  Thoughts?

**Comment by JeremySkinner** [comment]
Yes that's a good idea...if we differentiate between RuleFor(x => x).SetValidator and RuleFor(x => x.Property).SetValidator() then this is not as much of a breaking change, and makes much more sense in the context. I actually prefer this to having a new method like IncludeValidator

I'd be happy to include that change. I'm away for the weekend, but if you wanted to do a quick spike to put together some code for working out if an expression tree is in the form of x => x, then I'll take a look at getting it in next week. 

Jeremy

**Comment by waynebrantley** [comment]
In your code, you reference these expressions in a couple of ways.
So, here are two ways to determine the form.

Given `Expression<Func<T, TProperty>> expression`
If expression is of form `x => x.SomeProp` then
    `expression.Body.NodeType==MemberAccess`
    and
    `(expression.Body as MemberExpression) != null`

If expression is of form `x => x` then
    `expression.Body.NodeType==Parameter`
    and
    `(expression.Body as MemberExpression) == null`

Thanks

**Comment by JeremySkinner** [comment]
I've made the changes and checked them in...feel free to give it a build and let me know how it looks.

**Comment by waynebrantley** [comment]
Awesome.  Would it be possible for you to do a new nuget package.  If not, we have to get source, do a build - copy that dll around, etc.  It is painful.   Even if you want to mark it as a prerelease that would help.

If you can do this, I can easily test with my project and report back today.

**Comment by JeremySkinner** [comment]
It would be great if you could build from soruce for now....I don't want to do a nuget release until we're sure it's definitely working, and my build process isn't set up for doing prerelease packages. It should just be case of doing a pull and running build.cmd

**Comment by waynebrantley** [comment]
We constructed a workaround - it is ugly and 'hard coded' for our special situations, but it works.  I think if you had a 'prerelease process' that would be helpful for sure :-)

So yeah, i can wait.

**Comment by mrmashal** [comment]
Please excuse me if I'm being a distraction, but I think adding a magic bahavior like that might not be a good design decision.

I totally agree with @waynebrantley about the idea behind this thread and actually I found it while struggling with getting an intuitive design for my MVC application. I'm using my IoC container to instantiate FV validators. As a new user (and fan :) of FV, I was shocked when I understood that there is no `.SetValidator<TValidator>()` overload to be used in this scenario. The current way of doing `.SetValidator(new SomeValidator())`, IMO, is only reasonable when `SomeValidator` requires some input arguments in its constructor and there is no provider to automatically supply them for it.

But coming back to the issue, I think the concept of RuleSets can be generalized with some "Status" flag in the validation context which is shared among all the nested validations.
In my scenario, I'm differentiating between Create and Update requests using a `[CustomizeValidator(RuleSet = "Create")]` attribute for my retrieved ViewModel and it would be a significant advantage if the "Create" thing would propagate through all the nested validators, even if some validators in the middle don't care about it.

I have also tried to solve this by writing everything in the validator for the main ViewModel class like `RuleFor(x => x.Level1.Level2.…).NotEmpty()`, but it didn't give me the appropriate client-side validation.

So, to sum up, my points are as follows:
1. Why there is no `.SetValidator<TValidator>()`?
2. Please introduce a concept of "Status" or something in the validation context.
3. Is this a bug that a multi-level `RuleFor(x => x.Level1.Level2.…).NotEmpty()` has a different behavior than a direct `RuleFor(x => x.Level1).NotEmpty()`?

Maybe I should have opened other issues for these, but this is just how it came to my mind.
Thanks anyway ;)

**Comment by JeremySkinner** [comment]
@mrmashal,

1. SetValidator<TValidator>
There are a few reasons why this isn't practical. FluentValidation doesn't know anything about the ioc container that you're using, so would have no way to retrieve the validator instance. Also, that's really not the way am ioc container should be used....the container should construct the entire hierarchy, with dependencies specified through constructors....it's not typically a good idea for consuming code to call back into the container. For one, it's very opaque and hard to see what's actually happening, but also messes up object lifecycles. 

If you need to construct the child validate using you ioc container, then you should be treating FV like any other object graph...define the child validator as a dependency of your parent by requiring it in the constructor, and then pass this into SetValidator. 

2. Ruleset status
Not sure I understand what you meant by status? Maybe you could elaborate on that? As I mentioned in my second reply, if you want the rule set to always cascade down the you can implement this behaviour yourself with a custom ValidatorSelector. The selection mechanism for validators is pluggable, so if you don't like the defaults then you can swap it out using this extensibility point. You should be able to modify the code sample above with not much work, but let me know if you get stuck and I'd be happy to help. 

3. Multi-level rules
As far as the FV engine is concerned, there is no difference between single level rules and multi level rules. The behaviour you're seeing is one of the (many) limitations of integrating its mvc- you have to have a validator per type if you want client side rules (personally I don't use the mvc integration, it's a PITA) 

Hope that makes sense

**Comment by mrmashal** [comment]
About the first point, as you have mentioned in your [blog post](http://www.jeremyskinner.co.uk/2010/02/22/using-fluentvalidation-with-an-ioc-container/), we're setting up a singleton descendant of `FluentValidation.ValidatorFactoryBase` that is initialized at application startup with a dependency on the IoC container, and I think that's a common approach to manually make use of the container's registry. Let me ask you to reconsider this if you don't mind, since the support for IoC is so useful and practical. But I'm not going to divert the main topic of this thread more than this.

About the status concept, let me explain my use case: My form is submitted through a ViewModel containing nested objects of type `UploadedImage` each containing the temporary name of an asynchronously uploaded image and some dimensions for cropping it. upon creation of new entities, I need to make sure an image has been uploaded, but on later updates, there is no need to perform this check. I thought by using FV, I can make this logic more maintainable by using a single ViewModel with conditional validations. The required validations are very simple and I thought there would be no need to implement a custom server and client-side validation for them.

The problem is: the conditional validation for uploaded images should be done on a property `TempImageName` in the `UploadedImage` objects. It would be great if I could just decorate my ViewModel in the Create action of my controller with something like `[CustomizeValidator(Status = "Create")]` and this status would propagate to every sub-validator so it knows which rules to apply. I have understood that, currently, due to MVC limitations, I need to write more code for this seemingly simple problem but It's just a suggestion to improve the awesome library for more general use cases.

Thanks again

**Comment by JeremySkinner** [comment]
> and I think that's a common approach to manually make use of the container's registry. 

Yes, it is a common thing to do, but to create the root object of the hierarchy. That being said, there's nothing stopping you from directly calling into the container from inside the validator's constructor, but I'd ***highly*** suggest not doing this, and relying on constructor injection instead. This isn't something I'm going to change- sorry. 

Regarding the Status....it seems to me like this is exactly the scenario that you should use rulesets for. I don't think introducing another concept to do this is a good idea. As I mentioned before, adding a custom ValidatorSelector to allow downward propogration for rulesets is pretty simple to do...however, the CustomizeValidatorAttribute doesn't currently allow the use of custom Selectors (it's hard-coded to use the default ones). What I'd suggest is that I extend the CustomizeValidatorAttribute to allow the default validator selectors to be overriden via subclassing (example below). That way, you can then inject your own ValidatorSelector into the pipeline to customize the ruleset behaviour. Would that work for you?

```
public class MyCustomizeValidatorAttribute : CustomizeValidatorAttribute {
   protected override IValidatorSelector CreateRulesetValidatorSelector(string[] ruleSets) {
     return new IncludeChildValidatorSelector(ruleSets);
   }
}  
```

```
public ActionResult Foo([MyCustomizeValidator(RuleSets: "Names, Whatever")] Person person) {
 //...
}
```

**Comment by JeremySkinner** [comment]
Actually, it *is* already possible to use a custom selector with an overriden CustomizeValidatorAttribute....the interceptor API allows this already :smiley: 

Example:

```
public class MyCustomizeValidatorAttribute : CustomizeValidatorAttribute {

	public MyCustomizeValidatorAttribute() {
		Interceptor = typeof (MyCustomizeValidatorAttribute.InnerInterceptor);
	}

	public class InnerInterceptor : IValidatorInterceptor {
		public ValidationContext BeforeMvcValidation(ControllerContext controllerContext, ValidationContext validationContext) {
			if (validationContext.Selector is RulesetValidatorSelector) {
				return validationContext.Clone(selector: new IncludeChildValidatorSelector());
			}

			return validationContext;
		}

		public ValidationResult AfterMvcValidation(ControllerContext controllerContext, ValidationContext validationContext, ValidationResult result) {
			return result;
		}
	}
}

public class IncludeChildValidatorSelector : IValidatorSelector
{
	private IValidatorSelector _innerSelector;
	string[] _rulesetsToExecute;

	public IncludeChildValidatorSelector(params string[] ruleSets)
	{
		_rulesetsToExecute = ruleSets;
		_innerSelector = new RulesetValidatorSelector(ruleSets);
	}

	public bool CanExecute(IValidationRule rule, string propertyPath, ValidationContext context)
	{
		if (string.IsNullOrEmpty(rule.RuleSet) && _rulesetsToExecute.Length > 0)
		{
			bool isChildValidator = rule.Validators.FirstOrDefault() is ChildValidatorAdaptor;
			if (isChildValidator)
			{
				return true;
			}
		}

		return _innerSelector.CanExecute(rule, propertyPath, context);
	}
}
```

**Comment by JeremySkinner** [comment]
@waynebrantley Having slept on this...I'm starting to think that a top level "Include" method might actually be better than the RuleFor(x => x) approach. @mrmashal makes a good point that this is quite "magic" behaviour, so I'm now leaning towards "Include" for clarity. 

```
public MainValidator() {
   Include(new SharedValidator());
}
```

What do you think?

**Comment by waynebrantley** [comment]
@JeremySkinner That was my initial concern about the x=>x using it.  It is slick and cool, but it is not an easily discoverable feature.   As you saw, in my post above - I was leaning toward the Include() for that very reason.  So, that is good with me!

Just a note - your library is really fantastic and we use it in lots of places.  Really great work and I always appreciate your willing to listen, keep it updated, etc.  Thanks a bunch!

**Comment by mrmashal** [comment]
>  The only hiccup is getting access to the rulesets from inside the interceptor, but that's an easy change for me to make.

That would be nice of you. I guess that's the way to go. I'm not much familiar with the inner workings of the library and maybe that's the cause of all my confusion.

Anyway, Thanks for your time, Jeremy. Keep up the good work!

**Comment by JeremySkinner** [comment]
@waynebrantley  Thanks for the kind words! :smile:  I'll see if I can get this done this week, and put out a prerelease package for testing. 

@mrmashal ok, thanks- I'll try and get this change done at the same time. I'll let you know when it's done so you can give it a test. 

Thanks!

**Comment by mrmashal** [comment]
Something I remembered: The global hook for providing a default `ValidatorSelector` would also do the job for me.

**Comment by JeremySkinner** [comment]
Ah yes, I can add that in too

**Comment by JeremySkinner** [comment]
Sorry I haven't had the time to be able to get to this yet, but it's still on my radar.

**Comment by waynebrantley** [comment]
We did a hardcoded workaround that won't work long term, but for short term we are fine.   Thanks for keeping on your radar, we will take what we can get :-)

**Comment by waynebrantley** [comment]
@JeremySkinner Any idea when you might have a second to get this in for testing?

**Comment by JeremySkinner** [comment]
I've made the changes to add support for the Include statement and the interceptor changes the other day, and I've just pushed them to master. I've had no time to look at adding support for prerelease packages to the build process though, and probably won't for another few weeks, sorry.

**Comment by JeremySkinner** [comment]
Turns out I got some time this morning and added support for prerelease packages, so I've pushed up a 6.2.0-beta1 build. Please give this a try.

**Comment by waynebrantley** [comment]
@JeremySkinner LOL, thanks!   We will give it a try...

**Comment by waynebrantley** [comment]
@JeremySkinner The new pre-release build works as designed!  Nice job. 

The only thing I will mention that is odd (no necessarily related to this exact new code) is you have a unit test like this:
```
		[Fact]
		public void Ruleset_selection_should_cascade_downwards_with_when_setting_child_validator_using_include_statement() {
			var validator = new TestValidator3();
			var result = validator.Validate(new Person(), ruleSet:"Names");
			result.IsValid.ShouldBeFalse();
		}
```

The overload of Validate() you are calling there is not present in the IValidator interface, so that means we cannot use it.  Instead we create ValidationContext, etc.

Did not know if there was a reason not to, but it looks like IValidator interface should have an additional Validate overload to support the above?

**Comment by JeremySkinner** [comment]
That's an extension method that's just a shortcut to providing an explicit
validation context if you're working with to validator type directly.

If you're working with the IValidator interface you'll need to provide th
context directly- the methods exposed  on the interface should be kept to a
minimum.

**Comment by waynebrantley** [comment]
Ok, I thought you might say that.
As far as I am concerned, you can close this ticket and publish a new non-prerelease build.

**Comment by mrmashal** [comment]
Sorry if I'm late on this. I'm a little busy now. But I took a quick look at the codes and couldn't find out how to access the rulesets in my `BeforeMvcValidation`. I supposed you would add the rulesets to the `validationContext`. Is that not reasonable?

I also couldn't figure out how I should set my validator selector in `ValidatorOptions`.

**Comment by JeremySkinner** [comment]
No, they're not exposed on the context. They're on the selector- do a safe cast from IValidatorSelector to RulesetValidatorSelector and then you can retrieve them from there.

There are also the global options for setting the default selectors on the ValidatorOptions static class.

**Comment by JeremySkinner** [comment]
Great, thanks @waynebrantley - I'll try and get that done soon.

**Comment by hidegh** [comment]
**Auto setting validators for nested object:**

The IValidatableObject interface (if defined on model DTO's) is from MVC6 automatically bound to the DI context, so you're able to get anything for your services inside validation. Also IValidatableObject does validate also childrens (child objects) if they do provide an implementation for the IValidatableObject interface [i'm almost sure it does, but haven't used it for a while]. Unfortunatelly no "When" and "Unless" can be performed...

In case of FV using SetValidator(new ...) for child objects is annoying and also might become complex if dependencies are needed. While the MVC6 API layer does the 1st step, and uses DI to set up the main objects validator, it won't use DI and set validators for child properties.

*Solution (need to check):*
- create a basic validator which has a dependency context injected
- inside the basic constructor we should check the object type we validate and for each property that has the ValidatorAttribute assigned, we should get and set the given validator

By calling: RuleFor(x=>x.Child).When(x=>false) we could even bypass this default validator...
